src/typeEvaluator/typeEvaluate.ts

```diff
@@ -117,20 +119,103 @@ function handleDerefNode(node: DerefNode, scope: Scope): TypeNode {
 }
 
 function mapObjectSplat(
-  node: TypeNode,
+  node: ConcreteTypeNode,
   scope: Scope,
-  mapper: (field: Document | ObjectTypeNode) => void,
-) {
-  if (node.type === 'union') {
-    for (const scoped of node.of) {
-      mapObjectSplat(scoped, scope, mapper)
+  mapper: (attribute: ObjectAttribute) => ObjectAttribute,
+): TypeNode {
+  // if the node is not an object it means we are splating over a non-object, so we return unknown
+  if (node.type !== 'object') {
+    return {type: 'unknown'}
+  }
+
+  const attributes: Record<string, ObjectAttribute> = {}
+  for (const name in node.attributes) {
+    if (!node.attributes.hasOwnProperty(name)) {
+      continue
     }
+    attributes[name] = mapper(node.attributes[name])
   }
 
-  if (node.type === 'object') {
-    mapper(node)
+  if (node.rest !== undefined) {
+    // Rest is either an object, inline, or unknown - so mapping the concrete value here and returning it is just to resolve the inline
+    const concreteRest = mapConcrete(node.rest, scope, (rest) => rest)
+
+    // if the rest is unknown the entire object is unknown
+    if (concreteRest.type === 'unknown') {
+      return {type: 'unknown'}
+    }
+    if (concreteRest.type !== 'object') {
+      return {type: 'null'}
+    }
+    for (const name in concreteRest.attributes) {
+      // eslint-disable-next-line
+      if (!concreteRest.attributes.hasOwnProperty(name)) {
+        continue
+      }
+      attributes[name] = mapper(concreteRest.attributes[name])
+    }
   }
+  return {type: 'object', attributes}
 }
+
+function handleObjectSplatNode(attr: ObjectSplatNode, scope: Scope): TypeNode {
+  const value = walk({node: attr.value, scope})
+  $trace('object.splat.value %O', value)
+  return mapConcrete(value, scope, (node) => {
+    const attributes: Record<string, ObjectAttribute> = {}
+    const mapped = mapObjectSplat(node, scope, (attribute) => attribute)
+    if (mapped.type != 'object') {
+      return mapped
+    }
+    for (const name in mapped.attributes) {
+      if (!mapped.attributes.hasOwnProperty(name)) {
+        continue
+      }
+      attributes[name] = mapped.attributes[name]
+    }
+    return {type: 'object', attributes}
```

Comment:
This code has just basically takes a `type: "object"` and then makes a copy of it without `rest` or `dereferencesTo`. That doesn't look right to me.